### PR TITLE
enum types in daml-lf

### LINKED ISSUE #105: add enum types to DAML-LF
variants are kinda icky in many situations. i propose we add a new datatype definition along with `record` and `variant`, `enum`:

```
enum T = V_1 | ... | V_n
```

which carves out explicitly variants with no arguments. this is worth doing because we can encode as strings, which is _a lot_ more compact and intuitive both on the ledger API and in json etc.

### PR DESCRIPTION
**Pull request by bitonic**
fixes #105 

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala [hunk 1]
```diff
@@ -36,20 +40,23 @@ object DefDataType {
...
 }
 
 object DataType {
 
-  /** Alias for application to [[FieldWithType]]. */
-  type FWT = DataType[FieldWithType, FieldWithType]
+  /** Alias for application to [[Type]]. Note that FWT stands for "Field with
+    * type", because before we parametrized over both the field and the type,
+    * while now we only parametrize over the type.
+    */
```
Comment (S11001001) [inline daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala]: The original reason for letting you parameterize the strings too was so you could have a separate type for pre-name-mangled field names and post-name-mangled field names (i.e. Scala identifiers); if you don’t think that is a sufficiently interesting use case then this is fine.

File: language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/CodeGen.scala [hunk 2]
```diff
@@ -241,22 +242,21 @@ object CodeGen {
...
+                packageId,
+                qualName copy (name =
+                  DottedName.assertFromSegments(qualName.name.segments.slowSnoc(vn).toSeq)))
+              val key = (syntheticRecord, vTypeVars.toList)
+              typeVarDelegate(vt)
+                .filter((_: Identifier) == syntheticRecord)
+                .flatMap(_ => recordMap get key)
+                .cata(nr => (Set(key), (vn, -\/(nr.fields.toList))), (noDeletion, (vn, \/-(vt))))
           }
+          (deleted, ScopedDataType(ident, vTypeVars, Variant(sdt)))
```
Comment (S11001001) [inline language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/CodeGen.scala]: I’d rather keep the traverse layers; this reversion used to be how it was but it caused confusing flexibility problems with the original introduction of `ScopedDataType`. One of the many times I've been punished for ignoring abstractions I'm aware of in favor of "intuitive" inlined code. I would be ok with this if I believed that the `interface` library was not going to have any design changes in the future :)

Comment (S11001001) [reply language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/CodeGen.scala]: Eh, I guess the choice of taking the `(String,)` out of the tparam makes that impossible, never mind

File: daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala (hunk 1 above)
Comment (bitonic) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala]: it's not that i do not find the case sufficiently interesting, is that this abstraction does not work anymore where there are datatypes where you have _only_ labels (enums).

i think if we want to accomodate such definitions this change is the best compromise.

Comment (S11001001) [reply daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala]: I see; as long as we're clear that "not every label has an associated type" then that seems fine.

We could also introduce a third type parameter for enums...it's traditional to evolve to the final form of `Trifunctor` right? ;)

**Pull request closed without merging**